### PR TITLE
Modify documentation to be more clear about resolvers being static

### PR DESCRIPTION
**Pull request by allardhoeve**
See #951

### CODE REVIEW

File: docs/execution/dataloader.rst [hunk 1]
```diff
@@ -111,8 +111,10 @@ leaner code and at most 4 database requests, and possibly fewer if there are cac
         best_friend = graphene.Field(lambda: User)
         friends = graphene.List(lambda: User)
 
-        def resolve_best_friend(self, info):
-            return user_loader.load(self.best_friend_id)
+        @staticmethod
+        def resolve_best_friend(root, info):
```
Comment (jkimbo) [inline docs/execution/dataloader.rst]:
```suggestion
        def resolve_best_friend(user, info):
```

`root` has a specific meaning in graphql (it’s the value that is passed into the root of the tree) and so anything that is it the `Query` object should use variable names that indicate what the variable is, so `user` in this case.

File: docs/execution/dataloader.rst [hunk 2]
```diff
@@ -111,8 +111,10 @@ leaner code and at most 4 database requests, and possibly fewer if there are cac
...
+            return user_loader.load(root.best_friend_id)
 
-        def resolve_friends(self, info):
-            return user_loader.load_many(self.friend_ids)
+        @staticmethod
+        def resolve_friends(root, info):
```
Comment (jkimbo) [inline docs/execution/dataloader.rst]:
```suggestion
        def resolve_friends(user, info):
```

File: docs/relay/connection.rst [hunk 3]
```diff
@@ -41,5 +41,6 @@ that implements ``Node`` will have a default Connection.
         name = graphene.String()
         ships = relay.ConnectionField(ShipConnection)
 
-        def resolve_ships(self, info):
+        @staticmethod
+        def resolve_ships(root, info):
```
Comment (jkimbo) [inline docs/relay/connection.rst]:
```suggestion
        def resolve_ships(faction, info):
```

**Comment by allardhoeve** [comment]
> @allardhoeve decorating resolvers with @staticmethod is not necessary but thanks for the change from self

Good suggestions, I'll apply them next week and then re-request the merge.

I disagree about the necessity of `@staticmethod` though. I understand that you unbind the method in the resolver code, so technically for GraphQL it is not needed. But there are other things to consider besides this.

1. IDEs such as PyCharm will warn that the method, because it is still a method even though GraphQL unbinds it, should be marked as such.
2. These IDEs will also warn you that coding standards indicate that the first argument to a method is normally `self`.

Of course you can ignore these standards, but they are confusing to code reviewers and to programmers new to GraphQL, such as myself.

Lastly the methods are no longer callable from the outside of GraphQL without also using the unbind magic, because the signature no longer matches a method, because the `self` argument is not in the signature. This would prevent you from calling the resolver properly, for example in unittests.

I feel that the documentation for GraphQL should be as clear as possible and should reflect this rather odd unbinding of the resolver well, so new programmers are sure about the interface.

You can of course always leave the decorator off you own functions if you want. To me this is a documentation issue.

**Comment by ProjectCheshire** [comment]
I disagree on needing static also because resolvers do not need to be defined next to the fields, but may be defined outside of, and may very well be for more complex ones. 

In that case, adding staticmethod makes it confusing. And doing it only for resolvers when defined with the field makes it inconsistent and confusing.

**Comment by allardhoeve** [comment]
Indeed, not every resolver needs to be a staticmethod. If you use an outside function as a resolver, then no, you definitely shouldn't use a decorator. But as defined in a class, they probably should be marked as such, as they are essentially functions after becoming unbound by GraphQL.

My request is not at all to force everyone to use static method decorators everywhere, but just that we improve the documentation on resolvers better, because they are inconsistently documented currently.

My suggestion is still that for resolvers that are defined on NodeType objects, we still document them as static methods, as they are. And the part in the documentation on resolvers is easily missed.

But the situation improves with changing `self` with the subject of the function such as `ship`, thanks!

**Comment by phalt** [comment]
Is it possible to add a documentation note about this issue with some IDEs and leave the default examples without? Either way, we should progress this PR or close it.

**Pull request closed without merging**